velox/vector/arrow/Bridge.cpp

```diff
@@ -631,6 +703,71 @@ void exportValues(
   holder.setBuffer(1, values);
 }
 
+void exportViews(
+    const FlatVector<StringView>& vec,
+    const Selection& rows,
+    ArrowArray& out,
+    memory::MemoryPool* pool,
+    VeloxToArrowBridgeHolder& holder) {
+  auto stringBuffers = vec.stringBuffers();
+  size_t numStringBuffers = stringBuffers.size();
+  size_t numBuffers = 3 +
+      numStringBuffers; // nulls, values, stringBuffers, variadic_buffer_sizes
+
+  // resize and reassign holder buffers
+  holder.resizeBuffers(numBuffers);
+  out.buffers = holder.getArrowBuffers();
+  // cache for fast [buffer-idx, buffer-offset] calculation
+  std::vector<int32_t> stringBufferVec(numStringBuffers);
+
+  BufferPtr variadicBufferSizes =
+      AlignedBuffer::allocate<size_t>(numStringBuffers, pool);
+  auto rawVariadicBufferSizes = variadicBufferSizes->asMutable<uint64_t>();
+  for (int32_t idx = 0; idx < numStringBuffers; ++idx) {
+    rawVariadicBufferSizes[idx] = stringBuffers[idx]->size();
+    holder.setBuffer(2 + idx, stringBuffers[idx]);
+    stringBufferVec[idx] = idx;
+  }
+  holder.setBuffer(numBuffers - 1, variadicBufferSizes);
+  out.n_buffers = numBuffers;
+
+  // Sorting cache for fast binary search of [4-byte buffer-idx, 4-byte
```

Comment:
I think I'm missing something here. Could you add some more comments on the headers of this function to explain how this mapping works?

Why can't we just map the Velox string buffers to Arrow string view buffers? Not sure I follow what the code is doing.

Reply:
Sure, I'll add some comments to the headers. For reference, the export of Velox StringView to Arrow StringView requires searching for the 8-byte pointer within the `FlatVector::stringBuffers_` (`vector<BufferPtr>`) in order to obtain the `4-byte index` and `4-byte offset`. Instead of doing a linear search(n.m), I have made the change to sort and do a binary search (m.logm + n.logm) since m is expected to be small.

Reply:
I suspect in most cases, Velox will generate FlatBuffers with a single string buffer. To optimize for this base case, could you organize the code in a such as to provide a fast path implementation for the common single-buffer case, and a general (but slower) for the multiple cases where you need to build the cache, sort, then do the binary search?